R/interpolate.R

```diff
@@ -0,0 +1,359 @@
+#'                   value_col = "y", extrapolate_vals = c(1:10),
+#'                   method = "linear", n_groups_fit = 3)
+#'
+#' @export
+extrapolate <- function(dt, id_cols, extrapolate_col, value_col,
+                        extrapolate_vals, method, n_groups_fit) {
+
+  dt <- copy(dt)
+
+  # validate ----------------------------------------------------------------
+
+  # check `id_cols`
+  assertthat::assert_that(is.character(id_cols))
+
+  # check `dt`
+  assertable::assert_colnames(
+    dt, c(id_cols, value_col), only_colnames = F, quiet = T
+  )
+
+  # check `extrapolate_col`
+  assertthat::assert_that(
+    assertthat::is.string(extrapolate_col),
+    extrapolate_col %in% id_cols,
+    is.numeric(dt[, get(extrapolate_col)]),
+    msg = paste0("`extrapolate_col` must be an entry of `id_cols` and ",
+                 "represent a numeric column of `dt`.")
+  )
+
+  # check `value_col`
+  assertthat::assert_that(
+    assertthat::is.string(value_col),
+    value_col %in% names(dt),
+    is.numeric(dt[, get(value_col)]),
+    msg = "`value_col` must represent a numeric column of `dt`."
+  )
+
+  # check `extrapolate_vals`
+  assertthat::assert_that(is.numeric(extrapolate_vals))
+  lb <- min(dt[, get(extrapolate_col)])
+  ub <- max(dt[, get(extrapolate_col)])
+  assertthat::assert_that(
+    min(extrapolate_vals) < lb | max(extrapolate_vals) > ub,
+    msg = "`extrapolate_vals` are not outside of the range of the data."
+  )
+
+  # check `method`
+  methods <- c("linear", "rate_of_change", "uniform")
+  assertthat::assert_that(
+    assertthat::is.string(method),
+    method %in% methods,
+    msg = paste0("`method` is '", method, "' but must be one of: ",
+                 paste(methods, collapse = ", "))
+  )
+
+  # check `n_groups_fit`
+  assertthat::assert_that(assertthat::is.count(n_groups_fit))
+
+  # expand ------------------------------------------------------------------
+
+  # fill in NAs
+  expand_cols <- list(temp = extrapolate_vals)
+  names(expand_cols) <- extrapolate_col
+  dt <- expand(dt, expand_cols = expand_cols, id_cols = id_cols)
+
+  # extrapolate -------------------------------------------------------------
+
+  dt <- dt[order(get(extrapolate_col))]
+
+  dt <- rbindlist(
+    lapply(
+      split(dt, by = setdiff(id_cols, extrapolate_col)),
+      FUN = function(d) {
+
+        setnames(d, c(extrapolate_col, value_col), c("x", "y"))
+        # TODO: make this extrapolation work backwards as well
+        # subset to `n_groups_fit` groups
+        fit_data <- copy(d)
+        fit_data <- fit_data[!is.na(y)]
+        fit_data[, i := .I]
+        max_i <- max(fit_data$i)
+        fit_data <- fit_data[i %in% (max_i - n_groups_fit + 1):max_i]
+
+        if (method == "linear") {
+
+          # fit and predict linear model
+          fit <- stats::lm(data = fit_data, y ~ x)
+          d$y <- stats::predict(fit, newdata = d)
+
+        } else if (method == "rate_of_change") {
```

Comment:
Should this be equivalent to roc if `y` is log transformed before hand?

Reply:
I'm not sure I understand your question.

Reply:
Sorry I meant is this equivalent to the linear extrapolation when the y values are log transformed before using this function.

Similar to how log transforming the inputs prior to linear interpolation I think is equivalent to using the growth rate for interpolation. 

I first saw using growth rates for interpolation/extrapolation in this paper and is what I would expect our method to do. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4798897/#S6title